### PR TITLE
Serializable ArrayType Optimization

### PR DESCRIPTION
**Pull request by kubiak-jpl**
| | |
|:---|:---|
|**_Related Issue(s)_**| https://github.com/nasa/fprime/issues/4848,  |
|**_Has Unit Tests (y/n)_**| Updated existing unit tests |
|**_Documentation Included (y/n)_**| no new documentation |

---
## Change Description

- Adds an optimization in ArrayType when handing NumericalArray instances
  - Decreases the memory needed for each entry
  - Decreases the time to deserialize an entry
- Change json format of ArrayTypes to encode the member type information once, instead of in each entry
- Updates the Data Product decoder script to parse ArrayRecords as ArrayType structures
  - ArrayRecord json format now matches the format of arrays inside serializables
  - Speed up decoding of ArrayRecord entries

This is a breaking change for any scripts using the data produce parsing framework or working with and array serializable types

This builds of work by @olabie2 here https://github.com/nasa/fprime-gds/pull/289

## Rationale

Address memory usage for large arrays
https://github.com/nasa/fprime/issues/4848

Decrease decoding speed for large numerical arrays
https://github.com/nasa/fprime/issues/4784

## Testing/Review Recommendations

I've tested the current state of this PR on the unit tests and the data product parsing scripts, but it still needs wider testing on a GDS instance. I also need to test with arrays of structures that don't hit the optimizations here

## Open Questions

- Capitalized record dictionary keys ("Data", "Record") vs. lower case serializable keys ("size", "value", "values")
  - Should this inconsistency be addressed?
- Should the special optimization cases be handled by a subtype of ArrayType or through if else statements?

### CODE REVIEW

**Comment by kubiak-jpl** [comment]
## Profiling notes

#### Memory consumption
In the pre-test, at 10 MB data product used 2 GB memory for the image array storage. A growth factor of x200
```
../../lib/fprime-cameras/scripts/fprime-image-dp.py -d ../../build-artifacts/voxl2/NavCC/dict/NavCCTopologyDictionary.json -o out -b Dp_269484032_1773437517_00632834.fdp 
[ Top 10 ]
/tmp/fprime-gds/src/fprime_gds/common/dp/decoder.py:132: size=1909 MiB, count=20791269, average=96 B
...
```

Now a 50 MB data product uses ~400 MB. A growth factor of x8

```
ls -lh Dp_269488128_1773440401_00383442.fdp
-rw-r--r-- 1 kubiak kubiak 50M Mar 13 15:20 Dp_269488128_1773440401_00383442.fdp

../../lib/fprime-cameras/scripts/fprime-image-dp.py -d ../../build-artifacts/voxl2/NavCC/dict/NavCCTopologyDictionary.json -o out -b Dp_269488128_1773440401_00383442.fdp 
[ Top 10 ]
/home/kubiak/source/fprime-gds-kubiak/src/fprime_gds/common/models/serialize/array_type.py:110: size=398 MiB, count=214, average=1902 KiB
...
```

#### Runtime
In the pre-test a 12 MB data product and image writing script took 45 seconds to parse the data product
```
python3 -m cProfile  ./fprime-image-dp.py -d ../build-artifacts/voxl2/Voxl2Cams/dict/Voxl2CamsTopologyDictionary.json -b ../Dp_268455936_1772217218_00198277.fdp                                  23:13:03 [1319/1925]
         208448803 function calls (208331490 primitive calls) in 45.604 seconds                                   
                                                                                                                  
   Ordered by: cumulative time                                                                                    
                                                                                                                  
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)                                                                                                                                                              
    183/1    0.001    0.000   45.604   45.604 {built-in method builtins.exec}                                     
        1    0.367    0.367   45.604   45.604 fprime-image-dp.py:1(<module>)                                                                                                                                                         
        1    0.001    0.001   45.207   45.207 fprime-image-dp.py:15(main)                                         
        1    0.001    0.001   44.610   44.610 decoder.py:200(decode)                                             
      120    4.580    0.038   44.604    0.372 decoder.py:130(decode_record)                                       
 11550810   10.936    0.000   29.903    0.000 decoder.py:160(read_element)                                        
 11551428    3.268    0.000    9.079    0.000 type_base.py:98(to_jsonable)                                        
 23103447    5.305    0.000    6.493    0.000 numerical_types.py:35(getSize)                                   
 11551207    2.210    0.000    5.479    0.000 numerical_types.py:40(getMaxSize)            
 11551428    1.955    0.000    5.207    0.000 {built-in method builtins.repr}                                                                                                                                                        
 11551392    3.110    0.000    4.686    0.000 numerical_types.py:57(deserialize)                                                                                                                                                     
 11550933    3.652    0.000    3.652    0.000 {method 'tell' of '_io.BufferedReader' objects}            
 11551430    1.961    0.000    3.251    0.000 type_base.py:45(__repr__)                                           
 11551477    1.290    0.000    1.290    0.000 {method 'replace' of 'str' objects}                                 
 23101780    1.189    0.000    1.189    0.000 numerical_types.py:177(get_bits)                                    
 11551136    1.115    0.000    1.115    0.000 {method 'read' of '_io.BufferedReader' objects}              
 11557559    1.040    0.000    1.040    0.000 {method 'append' of 'list' objects}                   
 11551544    0.979    0.000    0.979    0.000 {built-in method _struct.unpack_from}                        
 11551818    0.812    0.000    0.812    0.000 type_base.py:63(__init__)                                    
 11551616    0.605    0.000    0.605    0.000 type_base.py:87(val)                                  
 11550817    0.597    0.000    0.597    0.000 numerical_types.py:182(get_serialize_format)
```

Now a 50 MB data product and image writing script took 1.38 seconds, with about half of that begin writing images to disk
```
python3 -m cProfile ../../lib/fprime-cameras/scripts/fprime-image-dp.py -d ../../build-artifacts/voxl2/NavCC/dict/NavCCTop
ologyDictionary.json -o out -b Dp_269488128_1773440401_00383442.fdp                                                                                                                                                                  
         1082383 function calls (791695 primitive calls) in 1.383 seconds                                  
                                                                                                                  
   Ordered by: cumulative time                                                                                    
                                                                                                                  
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)                                
    183/1    0.001    0.000    1.383    1.383 {built-in method builtins.exec}                 
        1    0.029    0.029    1.383    1.383 fprime-image-dp.py:1(<module>)                                                                                                                                                                 1    0.002    0.002    1.326    1.326 fprime-image-dp.py:15(main)                                                                                                                                                            
      106    0.722    0.007    0.722    0.007 {imwrite}                                                           
        1    0.032    0.032    0.299    0.299 decoder.py:200(decode)                                    
      424    0.001    0.000    0.247    0.001 decoder.py:131(decode_record)                        
      424    0.001    0.000    0.219    0.001 decoder.py:161(read_element)                          
       29    0.001    0.000    0.202    0.007 __init__.py:1(<module>)                               
      107    0.078    0.001    0.131    0.001 array_type.py:140(deserialize)                           
      106    0.018    0.000    0.117    0.001 __init__.py:120(dump) 
```

**Comment by kubiak-jpl** [comment]
## JSON structure

#### Serializable array
From
```
    "UserData": {                   
      "format": "{}",               
      "description": "User-configurable data",
      "name": "UserData",           
      "type": "UserData",           
      "size": 32,                   
      "values": [                   
        {                           
          "value": 0,               
          "type": "U8"              
        },                          
        {                           
          "value": 0,               
          "type": "U8"              
        },  
...
        {
          "value": 0,
          "type": "U8"
        }
      ]
    },
```

To
```
    "UserData": {
      "format": "{}",
      "description": "User-configurable data",
      "name": "UserData",
      "type": "UserData",
      "size": 32,
      "value_type": "U8",
      "values": [
        0,
        0,
        0,
        0,
        0,
...
        0,
        0
      ]
    },
```

#### ArrayRecord

From
```
    {     
      "Record": {
        "record_id": 269484035,
        "record_name": "NavCC.ov7251Cam.ImageRecord",
        "record_type_name": "U8Type",
        "record_type": "<class 'fprime_gds.common.models.serialize.numerical_types.U8Type'>",
        "is_array": true,
        "description": null
      },  
      "Size": 385024,
      "Data": [       
        {             
          "value": 38,
          "type": "U8"
        },            
        {             
          "value": 43,
          "type": "U8"
        },            
        {             
          "value": 40,
          "type": "U8"
        }, 
...
        {             
          "value": 0, 
          "type": "U8"
        }             
      ]               
    }, 
```

To
```
    {        
      "Record": {
        "record_id": 268455939,
        "record_name": "Voxl2.ov7251Cam.ImageRecord",
        "record_type_name": "U8Type",
        "record_type": "<class 'fprime_gds.common.models.serialize.numerical_types.U8Type'>",
        "is_array": true,
        "description": null
      },     
      "Data": {
        "name": "Voxl2.ov7251Cam.ImageRecord",
        "type": "Voxl2.ov7251Cam.ImageRecord",
        "size": 385024,
        "format": "{}",
        "value_type": "U8",
        "values": [
          11,
          11,
          12,
...
          0,             
          0,               
          0
        ]      
      }                                       
    },
```

**Comment by kubiak-jpl** [comment]
Converting to a draft because I found an issue with the current implementation. The `ArrayType.construct_type` line creates conflicting types in the dictionary since they all share the same name, but have different sizes. This causes at assert like the following
```
    data = DataProductDecoder(dictionaries, args.bin_file).decode()
  File "/home/kubiak/source/fprime-gds-kubiak/src/fprime_gds/common/dp/decoder.py", line 235, in decode
    record = self.decode_record(f, record_id)
  File "/home/kubiak/source/fprime-gds-kubiak/src/fprime_gds/common/dp/decoder.py", line 188, in decode_record
    element_array_type = ArrayType.construct_type(
        record_template.get_name(), record_type, array_size, "{}"
    )
  File "/home/kubiak/source/fprime-gds-kubiak/src/fprime_gds/common/models/serialize/array_type.py", line 39, in construct_type
    return DictionaryType.construct_type(
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
        cls, name, MEMBER_TYPE=member_type, LENGTH=length, FORMAT=format, DEFAULT=default
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/kubiak/source/fprime-gds-kubiak/src/fprime_gds/common/models/serialize/type_base.py", line 145, in construct_type
    original_properties == class_properties
AssertionError: Different class properties specified
```

**Comment by kubiak-jpl** [comment]
@thomas-bc Thoughts on this assertion?

One solution could be to name the new array type that gets created with the length appended. Something like
```
            element_array_type = ArrayType.construct_type(                  
                f'{record_template.get_name()}_{array_size}', record_type, array_size, "{}"                        
            )
```

**Comment by thomas-bc** [comment]
> One solution could be to name the new array type that gets created with the length appended.

Yes, this is exactly what is done in other parts of the code (can't remember where but a grep should find it). Let's do that

**Comment by thomas-bc** [comment]
@kubiak-jpl

**Comment by kubiak-jpl** [comment]
I'm able to run the Ref deployment with this patch and can see array types in the events and warnings in the GDS

Screenshot 2026-04-15 at 10 20 14 AM
Screenshot 2026-04-15 at 10 20 34 AM

I think this PR is good to go now

**Comment by thomas-bc** [comment]
I've opened nasa/fprime#5032 to trigger the full suite of CI tests running in F´ core with these changes. Through the following branch https://github.com/nasa/fprime-gds/tree/pr-5032 which is a clone of this PR here

**Comment by kubiak-jpl** [comment]
@thomas-bc @LeStarch 

Last week we discussed sending commands to the Ref deployment to test some edge cases in these changes. I tested these and it worked just fine, but I also don't think it was a test of my changes. It seems like command never actually creates an array type, just a JSON string with the command arguments that's turned into a command packet type.

I did run another test with the `typeDemo.GLUTTON_OF_CHOICE` command which does a good job of covering arrays of structs and arrays of numerical types and it runs, along with it's type output just fine.

Screenshot 2026-04-21 at 10 50 34 AM
Screenshot 2026-04-21 at 10 50 54 AM

I added some debug output within the GDS array_type and I can see that these commands trigger both the array of structs and array of numerical values paths within the code. I think this is good to go

```diff
diff --git a/src/fprime_gds/common/models/serialize/array_type.py b/src/fprime_gds/common/models/serialize/array_type.py
index 04f8e85..06729d9 100644
--- a/src/fprime_gds/common/models/serialize/array_type.py
+++ b/src/fprime_gds/common/models/serialize/array_type.py
@@ -64,8 +64,10 @@ class ArrayType(DictionaryType):
         if self._val is None:
             return None
         elif self._is_numerical_array():
+            print(f"[{self.__class__.__name__}].val JSON Numerical Array")
             return list(self._val)
         else:
+            print(f"[{self.__class__.__name__}].val JSON Type Array")
             return [item.val for item in self._val]
 
     @property
@@ -78,9 +80,11 @@ class ArrayType(DictionaryType):
         """
         result = []
         if self._is_numerical_array():
+            print(f"[{self.__class__.__name__}].formatted_val JSON Numerical Array")
             for item in self._val:
                 result.append(self.FORMAT.format(item))
         else:
+            print(f"[{self.__class__.__name__}].formatted_val JSON Type Array")
             for item in self._val:
                 if hasattr(item, "formatted_val"):
                     result.append(item.formatted_val)
```

```
[Ref.TooManyChoices].val JSON Type Array                 
[Ref.ManyChoices].val JSON Type Array                  
[Ref.ManyChoices].val JSON Type Array                 
[Array_U8Type_2].val JSON Numerical Array              
[Svc.ComQueueDepth].val JSON Numerical Array          
[Svc.BuffQueueDepth].val JSON Numerical Array          
[Ref.TooManyChoices].val JSON Type Array              
[Ref.ManyChoices].val JSON Type Array                 
[Ref.ManyChoices].val JSON Type Array                  
[Array_U8Type_2].val JSON Numerical Array              
[Ref.TooManyChoices].val JSON Type Array              
[Ref.ManyChoices].val JSON Type Array                  
[Ref.ManyChoices].val JSON Type Array                 
[Array_U8Type_2].val JSON Numerical Array              
[Ref.TooManyChoices].val JSON Type Array              
[Ref.ManyChoices].val JSON Type Array                 
[Ref.ManyChoices].val JSON Type Array                  
[Array_U8Type_2].val JSON Numerical Array 
```

File: src/fprime_gds/common/models/serialize/array_type.py [hunk 1]
```diff
@@ -83,49 +98,87 @@ def val(self, val: list):
...
             "type": self.__class__.__name__,
             "size": self.LENGTH,
             "format": self.FORMAT,
-            "values": (
-                None
-                if self._val is None
-                else [member.to_jsonable() for member in self._val]
-            ),
+            "value_type": repr(self.MEMBER_TYPE()),
+            "values": vals,
```
Comment (thomas-bc) [inline src/fprime_gds/common/models/serialize/array_type.py:123]: My main concern was about potential side effects of changing `to_jsonable()`, but it turns out to be a moot point, because `to_jsonable()` is not relied on by the system to pass things around. In fact it seems that DataProduct / Parameter tools are the only part of the GDS that does use this method. 

Flask instead uses other JSON encodings for `*Data` objects that only rely on the `val` attribute on ArrayType, not its entire `to_jsonable` representation ... https://github.com/nasa/fprime-gds/blob/301b3dc418828e0826c5b57496eaa0f66ae9e031/src/fprime_gds/flask/json.py#L112-L118

File: src/fprime_gds/common/models/serialize/array_type.py [hunk 2]
```diff
@@ -83,49 +98,87 @@ def val(self, val: list):
...
+            array_format = f"{value_endian}{self.LENGTH}{value_format}"
+            return struct.pack(array_format, *self._val)
+        else:
+            return b"".join([item.serialize() for item in self._val])
 
     def deserialize(self, data, offset):
         """Deserialize the members of the array"""
-        values = []
-        for field_index in range(self.LENGTH):
+        if issubclass(self.MEMBER_TYPE, NumericalType) and self.LENGTH > 0:
```
Comment (thomas-bc) [inline src/fprime_gds/common/models/serialize/array_type.py]:
```suggestion
        if self._is_numerical_array() and self.LENGTH > 0:
```

**Review by thomas-bc — COMMENTED** [review]
Looks good to me! One quick nit, and I would still like a second pair of eyes on this

File: src/fprime_gds/common/models/serialize/array_type.py (hunk 1 above)
Comment (LeStarch) [reply src/fprime_gds/common/models/serialize/array_type.py:123]: I confirm

**Review by LeStarch — APPROVED** [review]

**Comment by LeStarch** [comment]
Approved because @thomas-bc was ready to approve modulo my confirmation of that one nit.
